hw/top_earlgrey/syn/asic.constraints.sdc

```diff
@@ -162,9 +162,20 @@ set SPI_DEV_TCK 16.0
 create_clock -name SPI_DEV_CLK  -period ${SPI_DEV_TCK} [get_ports ${SPI_DEV_CLK_PIN}]
 set_clock_uncertainty ${SETUP_CLOCK_UNCERTAINTY} [get_clocks SPI_DEV_CLK]
 
+create_generated_clock -name SPI_DEV_IN_CLK -source SPI_DEV_CLK -divide_by 1 \
+    [get_pins top_earlgrey/u_spi_device/u_clk_spi_in_buf/clk_o]
+create_generated_clock -name SPI_DEV_OUT_CLK -source SPI_DEV_CLK -divide_by 1 \
+    -invert [get_pins top_earlgrey/u_spi_device/u_clk_spi_out_buf/clk_o]
+
+## TODO: Define SRAM clock, which is muxed clock. Do we need set_case_analysis?
```

Comment:
Are both clocks valid clocks for the SRAM? If so, I believe we do not have to set case analysis, since both modes should be timed using STA...

Reply:
I don't know SDC well. Could you let me know how I can describe below scenario?

1. If SRAM clock is IO_DIV4, then only check timing for SRAM + fwmode modules.
2. If SRAM clock is SCK, then ignore fwmode and check timing for other modules. 

I've skimmed the DC constraints manual but not yet found useful info to solve. I will take a look again later.

Reply:
@eunchan  Are you trying to stop the synthesizer from working on timing paths between the clocks, or optimizing the logic depending on the mode.    The first one is easy (keep the clocks exclusive) the second one can be solved in a few ways, but the ones I know of aren't pretty (case analysis, stoping clock propagation, disabling timing, etc).   We shouldn't do those unless there is a strong reason, as they can have undetected consequences.

Reply:
Yeah if multiple operating modes are involved it can get tricky in synthesis. I think it is best to have something relatively generic and simple here (if possible) and formulate precise timing constraints using MMMC in the backend.

As for your question, @eunchan, does it matter if certain unused logic modules are still getting timed in modes where they are not being used (e.g. fwmode during 2.)?
Because if not, keeping the clocks exclusive as @dsandrs mentioned should be good enough.

Reply:
Thanks for comment Don, I meant the second. I know this design is a little bit ugly, but to share same DPSRAM, this is inevitable unless reworking fwmode to use SCK for SRAM access. If we want to use SCK for SRAM access from HW side always, we can't send data at the very first in FW mode as no data can be read at the time.

If we discard FW mode and use SPI Flash mode always (to access internal content), then it will be easier.